### PR TITLE
New Resource: `azurerm_application_security_group`

### LINKED ISSUE #369: Feature Request: Application Security Groups
Hi, Following the announcements at Ignite it would be nice to explore the Application Security Groups functionality which is now in preview via terraform. It seems to make security groups more like AWS and infinitely more friendly than they currently are.

https://docs.microsoft.com/en-us/azure/virtual-network/create-network-security-group-preview

### PR DESCRIPTION
**Pull request by tombuildsstuff**
Tasks:

- [x] Upgrade to v12 of the Azure SDK for Go
- [x] Document that this is still in Public Preview
- [x] Fix the rebase issue in `azurerm/config.go`

Fixes #369

### CODE REVIEW

**Comment by tombuildsstuff** [comment]
Tests pass:

screen shot 2018-02-28 at 11 47 37

File: azurerm/config.go [hunk 1]
```diff
@@ -323,22 +324,15 @@ func getArmClient(c *authentication.Config) (*ArmClient, error) {
 		return keyVaultSpt, nil
 	})
 
-	csc := containerservice.NewContainerServicesClientWithBaseURI(endpoint, c.SubscriptionID)
```
Comment (katbyte) [inline azurerm/config.go]: I'm at a loss as to why this was removed as it seems unrelated?

Comment (tombuildsstuff) [reply azurerm/config.go]: So this is actually being registered twice in master at the moment (due to a previous merge conflict):

https://github.com/terraform-providers/terraform-provider-azurerm/blob/f76c05bc2cc04e0a1974d0182497abc07abeba18/azurerm/config.go#L326

https://github.com/terraform-providers/terraform-provider-azurerm/blob/f76c05bc2cc04e0a1974d0182497abc07abeba18/azurerm/config.go#L503

this consolidates this into one
